File: src/data/courses/dsa/trie/binary-trie/offline-xor-queries/code/java/OfflineXorQueries.java
import java.util.*;

class OfflineXorQueries {
    class TrieNode {
        TrieNode[] children = new TrieNode[2];
    }
    
    public int[] maximizeXor(int[] nums, int[][] queries) {
        Arrays.sort(nums);
        int[][] indexedQueries = new int[queries.length][3];
        for (int i = 0; i < queries.length; i++) {
            indexedQueries[i] = new int[]{queries[i][0], queries[i][1], i};
        }
        Arrays.sort(indexedQueries, (a, b) -> a[1] - b[1]);
        
        int[] result = new int[queries.length];
        TrieNode root = new TrieNode();
        int idx = 0;
        
        for (int[] query : indexedQueries) {
            int x = query[0], m = query[1], qIdx = query[2];
            while (idx < nums.length && nums[idx] <= m) {
                insert(root, nums[idx++]);
            }
            result[qIdx] = idx == 0 ? -1 : getMaxXor(root, x);
        }
        return result;
    }
    
    private void insert(TrieNode root, int num) {
        TrieNode node = root;
        for (int i = 31; i >= 0; i--) {
            int bit = (num >> i) & 1;
            if (node.children[bit] == null) node.children[bit] = new TrieNode();
            node = node.children[bit];
        }
    }
    
    private int getMaxXor(TrieNode root, int num) {
        TrieNode node = root;
        int xor = 0;
        for (int i = 31; i >= 0; i--) {
            int bit = (num >> i) & 1;
            int opposite = 1 - bit;
            if (node.children[opposite] != null) {
                xor |= (1 << i);
                node = node.children[opposite];
            } else {
                node = node.children[bit];
            }
        }
        return xor;
    }
}

File: src/data/courses/dsa/trie/binary-trie/offline-xor-queries/code/python/offline_xor_queries.py
class TrieNode:
    def __init__(self):
        self.children = [None, None]

class OfflineXorQueries:
    def insert(self, root: TrieNode, num: int) -> None:
        node = root
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            if not node.children[bit]:
                node.children[bit] = TrieNode()
            node = node.children[bit]
    
    def getMaxXor(self, root: TrieNode, num: int) -> int:
        node = root
        xor_val = 0
        for i in range(31, -1, -1):
            bit = (num >> i) & 1
            opposite = 1 - bit
            if node.children[opposite]:
                xor_val |= (1 << i)
                node = node.children[opposite]
            else:
                node = node.children[bit]
        return xor_val
    
    def maximizeXor(self, nums: list[int], queries: list[list[int]]) -> list[int]:
        nums.sort()
        indexed_queries = [(q[0], q[1], i) for i, q in enumerate(queries)]
        indexed_queries.sort(key=lambda x: x[1])
        
        result = [0] * len(queries)
        root = TrieNode()
        idx = 0
        
        for x, m, q_idx in indexed_queries:
            while idx < len(nums) and nums[idx] <= m:
                self.insert(root, nums[idx])
                idx += 1
            result[q_idx] = -1 if idx == 0 else self.getMaxXor(root, x)
        
        return result
